datumaro/util/scope.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (C) 2021 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from contextlib import ExitStack, contextmanager
+from functools import partial, wraps
+from typing import Any, Callable, ContextManager, Dict, Optional, Tuple, TypeVar
+import threading
+
+from attr import attrs
+
+from datumaro.util import optional_arg_decorator
+
+T = TypeVar('T')
+
+class Scope:
+    """
+    A context manager that allows to register error and exit callbacks.
+    """
+
+    _thread_locals = threading.local()
+
+    @attrs(auto_attribs=True)
+    class ExitHandler:
+        callback: Callable[[], Any]
+        ignore_errors: bool = True
+
+        def __exit__(self, exc_type=None, exc_value=None, exc_traceback=None):
+            try:
+                self.callback()
+            except Exception:
+                if not self.ignore_errors:
+                    raise
+
+    @attrs
+    class ErrorHandler(ExitHandler):
+        def __exit__(self, exc_type=None, exc_value=None, exc_traceback=None):
+            if exc_type:
+                return super().__exit__(exc_type=exc_type, exc_value=exc_value,
+                    exc_traceback=exc_traceback)
+
+
+    def __init__(self):
+        self._stack = ExitStack()
+        self.enabled = True
+
+    def on_error_do(self, callback: Callable,
+            *args, kwargs: Optional[Dict[str, Any]] = None,
+            ignore_errors: bool = False):
```

Comment:
```suggestion
            ignore_errors: bool = False) -> None:
```

(and ditto for the other `None`-returning functions)

Reply:
What is the point in this?

Reply:
To help typecheckers find errors.

Reply:
So you suggest to add it to all function without return value?

Reply:
I would say yes (if there are annotations on the function's arguments). Otherwise, the return type defaults to `Any`.

Reply:
I'm not convinced it is worthy investment of efforts in most cases. I'm open for a PR on this, but I don't think it is needed for the functions, which names are not suppose any return value.